File: ts/example/src/complex.ts
// XXX: 3 complex examples, but complexity like this should be design smells.


// BAD: Should redesign to achieve Command-Query Separation.
export interface ExampleBothCommandAndQuery {
  doSomething(foo: boolean, bar: number): number;
}


// BAD: Should separate into several small interfaces.
export interface ExampleManyMethods {
  foo(i: number): void;
  bar(): number;
  baz(j: number): number;
}


// BAD: Should use Promises if available.
export interface ExampleCallback {
  doSomethingAsync(callback: (x: string) => void): void;
}

File: swift/example/Sources/VanillaMockExample/Complex.swift
// XXX: 3 complex examples, but complexity like this should be design smells.


// BAD: Should redesign to achieve Command-Query Separation.
protocol ExampleBothCommandAndQueryProtocol {
    func doSomething(foo: Bool, bar: Int) -> Int
}


// BAD: Should separate into several small interfaces.
protocol ExampleManyMethodsProtocol {
    func foo(_ i: Int)
    func bar() -> Int
    func baz(_ j: Int) -> Int
}


// BAD: Should use Promises-like async operation abstractions if available.
protocol ExampleCallbackProtocol {
    func doSomethingAsync(callback: @escaping (String) -> Void)
}